From 28b61887b80d2c090a33c6d3fd8d8aaac8094ad1 Mon Sep 17 00:00:00 2001
From: Bowgo Tsai <bowgotsai@google.com>
Date: Fri, 28 Sep 2018 12:05:02 +0800
Subject: [PATCH 3/8] allow adb to remount symlink mount points

Currently `adb remount` won't remount symlink mount points.
In Android Generic System Image, there is a symlink
/product -> /system/product for devices with and without a physical
/product partition to work, respectively:

  - Mount product partition under /system/product via
    'mount /product' OR

  - Keep using /product -> /system/product symlink,
    when no product partition

Currently find_proc_mount() is seeking "/product" under /proc/mounts.
But the actual mount path is "/system/product" when GSI is used
on a device with product partition.

Bug: 111539442
Test: adb remount && touch /product/abc on both GSI and non-GSI

Change-Id: I8f15a67109d0a3f4ee18596ef7eb4280c5631b11
Merged-In: I8f15a67109d0a3f4ee18596ef7eb4280c5631b11
(cherry picked from commit 41649b871a21f3f7ced97b9c2a3491a959bc4dd5)

Former-commit-id: c44b0371e8fe872375fcd5b28fd967f06abfa93a
Former-commit-id: 00cacda4b4cc882211c13eb50fdb1a6f11f1f60a
Former-commit-id: 896765fcb4167c7a788c49f928894a00af1f76a1
---
 adb/remount_service.cpp | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/adb/remount_service.cpp b/adb/remount_service.cpp
index d679a6dff4..67aeafed2e 100644
--- a/adb/remount_service.cpp
+++ b/adb/remount_service.cpp
@@ -29,6 +29,7 @@
 
 #include <string>
 
+#include <android-base/file.h>
 #include <android-base/properties.h>
 
 #include "adb.h"
@@ -36,6 +37,8 @@
 #include "adb_utils.h"
 #include "fs_mgr.h"
 
+using android::base::Realpath;
+
 // Returns the device used to mount a directory in /proc/mounts.
 static std::string find_proc_mount(const char* dir) {
     std::unique_ptr<FILE, int(*)(FILE*)> fp(setmntent("/proc/mounts", "r"), endmntent);
@@ -43,9 +46,15 @@ static std::string find_proc_mount(const char* dir) {
         return "";
     }
 
+    // dir might be a symlink, e.g., /product -> /system/product in GSI.
+    std::string canonical_path;
+    if (!Realpath(dir, &canonical_path)) {
+        PLOG(ERROR) << "Realpath failed: " << dir;
+    }
+
     mntent* e;
     while ((e = getmntent(fp.get())) != nullptr) {
-        if (strcmp(dir, e->mnt_dir) == 0) {
+        if (canonical_path == e->mnt_dir) {
             return e->mnt_fsname;
         }
     }
-- 
2.17.1

